File: backend/middleware/logger.py
import time
from starlette.middleware.base import BaseHTTPMiddleware
from starlette.requests import Request
from starlette.responses import Response


class RequestLoggerMiddleware(BaseHTTPMiddleware):
    """Lightweight request logging middleware."""

    async def dispatch(self, request: Request, call_next) -> Response:
        start = time.time()
        path = request.url.path
        if request.url.query:
            path = f"{path}?{request.url.query}"

        response = await call_next(request)
        duration = int((time.time() - start) * 1000)
        status = response.status_code

        if status >= 500:
            status_text = f"\033[31m[{status}]\033[0m"
        elif status >= 400:
            status_text = f"\033[33m[{status}]\033[0m"
        elif status >= 200:
            status_text = f"\033[32m[{status}]\033[0m"
        else:
            status_text = f"[{status}]"

        print(
            f"[Server] {request.method} to {path} - "
            f"Status: {status_text} - Duration: {duration}ms"
        )
        return response

File: server/middleware/logger.ts
import { Request, Response, NextFunction } from "express";

/**
 * A lightweight request logging middleware.
 * Logs method, path, response status, and duration.
 */
export function requestLogger(req: Request, res: Response, next: NextFunction) {
  const start = Date.now();
  
  // Clean request path
  const path = req.originalUrl || req.url;

  // Once request is finished, log the details
  res.on("finish", () => {
    const duration = Date.now() - start;
    const status = res.statusCode;
    
    // Select color wrapper depending on status (helps log visibility in development)
    let statusText = `[${status}]`;
    if (status >= 500) {
      statusText = `\x1b[31m[${status}]\x1b[0m`; // Red
    } else if (status >= 400) {
      statusText = `\x1b[33m[${status}]\x1b[0m`; // Yellow
    } else if (status >= 200) {
      statusText = `\x1b[32m[${status}]\x1b[0m`; // Green
    }

    console.log(
      `[Server] ${req.method} to ${path} - Status: ${statusText} - Duration: ${duration}ms`
    );
  });

  next();
}
